File: bukkit/src/main/java/me/luizotavio/minecraft/world/chunkloader/SlimeChunkLoader.java
/*
 * MIT License
 *
 * Copyright (c) [2022] [LUIZ O. F. CORRÊA]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.luizotavio.minecraft.world.chunkloader;

import gnu.trove.map.hash.TLongObjectHashMap;
import me.luizotavio.minecraft.prototype.chunk.ProtoSlimeChunk;
import me.luizotavio.minecraft.world.CraftSlimeWorld;
import net.minecraft.server.v1_8_R3.*;
import org.bukkit.craftbukkit.v1_8_R3.util.LongHash;

import java.io.IOException;

/**
 * @author Luiz Otávio de Farias Corrêa
 * @since 13/08/2022
 */
public class SlimeChunkLoader implements IChunkLoader {

    private final TLongObjectHashMap<ProtoSlimeChunk> chunks;

    public SlimeChunkLoader(TLongObjectHashMap<ProtoSlimeChunk> chunks) {
        this.chunks = chunks;
    }

    @Override
    public Chunk a(World world, int i, int i1) throws IOException {
        return ((WorldServer) world).chunkProviderServer
            .emptyChunk;
    }

    /**
     * Save method to save each chunk to the file system.
     */
    @Override
    public void a(World world, Chunk chunk) throws IOException, ExceptionWorldConflict {
    }

    /**
     * Save method to save each chunk to the file system.
     */
    @Override
    public void b(World world, Chunk chunk) throws IOException {
    }

    @Override
    public void a() {
    }

    @Override
    public void b() {
    }

    /**
     * Loads all chunks from the file system.
     *
     * @param craftSlimeWorld The world to load the chunk into.
     */
    public void loadAll(CraftSlimeWorld craftSlimeWorld) {
        ChunkProviderServer chunkProviderServer = craftSlimeWorld.chunkProviderServer;

        for (ProtoSlimeChunk chunk : chunks.valueCollection()) {
            Chunk target = chunk.toChunk(craftSlimeWorld);

            chunkProviderServer.chunks.put(
                LongHash.toLong(target.locX, target.locZ),
                target
            );

            target.addEntities();
        }
    }
}

File: stream/src/main/kotlin/io/github/luizotavio/slimekorld/loader/SlimeChunkLoader.kt
package io.github.luizotavio.slimekorld.loader

import gnu.trove.map.hash.TLongObjectHashMap
import io.github.luizotavio.slimekorld.pojo.ProtoSlimeChunk
import io.github.luizotavio.slimekorld.stream.DefaultSlimeCodec
import net.minecraft.server.v1_8_R3.Chunk
import net.minecraft.server.v1_8_R3.IChunkLoader
import net.minecraft.server.v1_8_R3.World
import org.bukkit.craftbukkit.v1_8_R3.util.LongHash
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile

class SlimeChunkLoader(
    file: File
) : IChunkLoader {

    private val protoChunks: TLongObjectHashMap<ProtoSlimeChunk>
    private val loadedChunks: TLongObjectHashMap<Chunk> = TLongObjectHashMap()

    init {
        if (!file.exists()) {
            throw IOException("File not found: $file")
        }

        val randomAccessFile = RandomAccessFile(file, "r")

        protoChunks = DefaultSlimeCodec(randomAccessFile)
            .run { read().protoChunks }
    }

    /**
     * Converts the proto chunk at the specified coordinates
     * into a Minecraft chunk.
     *
     * @param world the world the chunk is in
     * @return the loaded chunk, or `null` if not populated
     */
    private fun loadProtoChunk(world: World, hash: Long): Chunk? {
        val proto: ProtoSlimeChunk = protoChunks.remove(hash) ?: return null

        val loaded: Chunk = proto.load(world)

        loadedChunks.put(hash, loaded)
        return loaded
    }

    /**
     * Loads the chunk at the specified chunk coordinates.
     *
     * @param world the world
     * @param chunkX the chunk x-coordinate
     * @param chunkZ the chunk z-coordinate
     * @return the loaded chunk, or `null` if couldn't load
     */
    override fun a(world: World, chunkX: Int, chunkZ: Int): Chunk? {
        val hash = LongHash.toLong(chunkX, chunkZ)
        var chunk = loadedChunks[hash]
        if (chunk == null) {
            chunk = loadProtoChunk(world, hash)
        }
        return chunk
    }

    override fun a(world: World?, chunk: Chunk?) {
        // NOOP: saveChunk
    }

    @Throws(IOException::class)
    override fun b(world: World?, chunk: Chunk?) {
        // NOOP: writeChunkToNBT
    }

    override fun a() {
        // NOOP: chunkTick
    }

    override fun b() {
        // NOOP: saveExtraData
    }
}